File: dotnet/src/MyDotey.SCF.Labeled/Labeled/DefaultLabeledConfigurationManager.cs
using System;
using System.Collections.Generic;
using System.Reflection;
using System.Linq;
using sType = System.Type;

using NLog;

namespace MyDotey.SCF.Labeled
{
    /**
     * @author koqizhao
     *
     * Jun 15, 2018
     */
    public class DefaultLabeledConfigurationManager : DefaultConfigurationManager, ILabeledConfigurationManager
    {
        private static readonly Logger Logger = LogManager.GetCurrentClassLogger(typeof(DefaultLabeledConfigurationManager));

        private MethodInfo NoLabelGetPropertyValueMethod = null;
        private MethodInfo LabeledGetPropertyValueMethod = null;

        public DefaultLabeledConfigurationManager(ConfigurationManagerConfig config)
            : base(config)
        {
            MethodInfo[] methods = GetType().GetMethods(BindingFlags.Instance | BindingFlags.NonPublic);
            NoLabelGetPropertyValueMethod = methods.Where(m => m.Name == "GetPropertyValue"
                && m.IsGenericMethod && m.GetGenericArguments().Length == 2
                && m.GetParameters().Length == 2).Single();

            methods = GetType().GetMethods(BindingFlags.Instance | BindingFlags.NonPublic);
            LabeledGetPropertyValueMethod = methods.Where(m => m.Name == "GetPropertyValue"
                && m.IsGenericMethod && m.GetGenericArguments().Length == 2
                && m.GetParameters().Length == 3).Single();
        }

        public override V GetPropertyValue<K, V>(PropertyConfig<K, V> propertyConfig)
        {
            if (propertyConfig == null)
                throw new ArgumentNullException("propertyConfig is null");

            if (!(propertyConfig.Key is ILabeledKey))
                return base.GetPropertyValue(propertyConfig);

            IPropertyConfig noLabelPropertyConfig =
                AbstractLabeledConfigurationSource.MakeNoLabelConfig(propertyConfig);
            for (PropertyLabels propertyLabels = ((ILabeledKey)propertyConfig.Key).Labels;
                propertyLabels != null; propertyLabels = propertyLabels.Alternative)
            {
                foreach (IConfigurationSource source in SortedSources.Values)
                {
                    if (!(source is ILabeledConfigurationSource) && propertyLabels.Labels.Count != 0)
                        continue;

                    V value = default(V);
                    if (propertyLabels.Labels.Count == 0)
                        value = (V)GetPropertyValue(source, noLabelPropertyConfig);
                    else
                        value = (V)GetPropertyValue(source, noLabelPropertyConfig, propertyLabels.Labels);

                    value = ApplyValueFilter(propertyConfig, value);

                    if (!object.Equals(value, default(V)))
                        return value;
                }
            }

            return propertyConfig.DefaultValue;
        }

        protected virtual V GetPropertyValue<K, V>(ILabeledConfigurationSource source,
            PropertyConfig<K, V> propertyConfig, ICollection<IPropertyLabel> labels)
        {
            V value = default(V);
            try
            {
                value = source.GetPropertyValue(propertyConfig, labels);
            }
            catch (Exception e)
            {
                string message = string.Format(
                        "error occurred when getting property value, ignore the source. source: {0}, propertyConfig: {1}",
                        source, propertyConfig);
                Logger.Error(e, message);
            }

            return value;
        }

        protected virtual object GetPropertyValue(object source, IPropertyConfig noLabelPropertyConfig)
        {
            MethodInfo genericMethod = NoLabelGetPropertyValueMethod.MakeGenericMethod(
                noLabelPropertyConfig.GetType().GetGenericArguments());
            return genericMethod.Invoke(this, new object[] { source, noLabelPropertyConfig });
        }

        protected virtual object GetPropertyValue(
            object source, IPropertyConfig noLabelPropertyConfig, object labels)
        {
            MethodInfo genericMethod = LabeledGetPropertyValueMethod.MakeGenericMethod(
                noLabelPropertyConfig.GetType().GetGenericArguments());
            return genericMethod.Invoke(this, new object[] { source, noLabelPropertyConfig, labels });
        }
    }
}

File: java/src/main/java/org/mydotey/scf/labeled/DefaultLabeledConfigurationManager.java
package org.mydotey.scf.labeled;

import java.util.Collection;
import java.util.Objects;

import org.mydotey.scf.ConfigurationManagerConfig;
import org.mydotey.scf.ConfigurationSource;
import org.mydotey.scf.DefaultConfigurationManager;
import org.mydotey.scf.PropertyConfig;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * @author koqizhao
 *
 * Jun 15, 2018
 */
public class DefaultLabeledConfigurationManager extends DefaultConfigurationManager
        implements LabeledConfigurationManager {

    private static final Logger LOGGER = LoggerFactory.getLogger(DefaultConfigurationManager.class);

    public DefaultLabeledConfigurationManager(ConfigurationManagerConfig config) {
        super(config);
    }

    @SuppressWarnings({ "unchecked", "rawtypes" })
    @Override
    protected <K, V> Tuple<V, ConfigurationSource> doGetPropertyValue(PropertyConfig<K, V> propertyConfig) {
        Objects.requireNonNull(propertyConfig, "propertyConfig is null");

        if (!(propertyConfig.getKey() instanceof LabeledKey))
            return super.doGetPropertyValue(propertyConfig);

        PropertyConfig<?, V> noLabelPropertyConfig = AbstractLabeledConfigurationSource
                .removeLabels((PropertyConfig) propertyConfig);
        for (PropertyLabels propertyLabels = ((LabeledKey) propertyConfig.getKey())
                .getLabels(); propertyLabels != null; propertyLabels = propertyLabels.getAlternative()) {

            for (ConfigurationSource source : getSortedSources().values()) {
                if (!(source instanceof LabeledConfigurationSource) && !propertyLabels.getLabels().isEmpty())
                    continue;

                V value = null;
                if (propertyLabels.getLabels().isEmpty())
                    value = getPropertyValue(source, noLabelPropertyConfig);
                else
                    value = getPropertyValue((LabeledConfigurationSource) source, noLabelPropertyConfig,
                            propertyLabels.getLabels());

                value = applyValueFilter(source, propertyConfig, value);

                if (value != null)
                    return new Tuple<V, ConfigurationSource>(value, source);
            }
        }

        return new Tuple<V, ConfigurationSource>(propertyConfig.getDefaultValue(), null);
    }

    protected <K, V> V getPropertyValue(LabeledConfigurationSource source, PropertyConfig<K, V> propertyConfig,
            Collection<PropertyLabel> labels) {
        V value = null;
        try {
            value = source.getPropertyValue(propertyConfig, labels);
        } catch (Exception e) {
            String message = String.format(
                    "error occurred when getting property value, ignore the source. source: %s, propertyConfig: %s",
                    source, propertyConfig);
            LOGGER.error(message, e);
        }

        return value;
    }

}
